### PR TITLE
Start and end of tournament by environment variable

### PR DESCRIPTION
**Pull request by dokato**
Addressed #25 & #24 .

### CODE REVIEW

File: server/botany/settings.py [hunk 1]
```diff
@@ -23,8 +23,13 @@
 BOTANY_NUM_ROUNDS = int(os.environ["BOTANY_NUM_ROUNDS"])
 BOTANY_OPCODE_LIMIT = int(os.environ["BOTANY_OPCODE_LIMIT"])
 
-# TODO read this from environment (and make optional)
-BOTANY_TOURNAMENT_CLOSE_AT = datetime(2018, 9, 18, 14, 30, tzinfo=bst)
+BOTANY_TOURNAMENT_START_AT = os.environ.get("BOTANY_TOURNAMENT_START_AT")
```
Comment (mjpieters) [inline server/botany/settings.py]: `os.environ` only contains strings. You'll need to parse the string into a `datetime` object here

File: server/botany/settings.py [hunk 2]
```diff
@@ -23,8 +23,13 @@
...
+if BOTANY_TOURNAMENT_START_AT is None:
+    BOTANY_TOURNAMENT_START_AT = datetime.now(timezone.utc) - timedelta(1)
+
+BOTANY_TOURNAMENT_CLOSE_AT = os.environ.get("BOTANY_TOURNAMENT_CLOSE_AT")
```
Comment (mjpieters) [inline server/botany/settings.py]: Ditto here.

File: server/botany/views.py [hunk 3]
```diff
@@ -27,14 +27,19 @@
     summary,
 )
 
+def tournament_state():
```
Comment (mjpieters) [inline server/botany/views.py]: I think this could be improved with a better name, like `tournament_is_open()`.

File: server/botany/views.py [hunk 4]
```diff
@@ -27,14 +27,19 @@
...
+    "Flag saying whether tournament is open (True) or closed (False)"
+    close_condition = datetime.now(timezone.utc) < settings.BOTANY_TOURNAMENT_CLOSE_AT
+    start_condition = datetime.now(timezone.utc) > settings.BOTANY_TOURNAMENT_START_AT
+    return close_condition and start_condition
```
Comment (mjpieters) [inline server/botany/views.py]: This now creates `datetime.now()` twice. That could be avoided with a chained comparison:

```
def tournament_is_open():
    """Has the tournament started and has not yet reached the closing date?"""
    start = settings.BOTANY_TOURNAMENT_START_AT
    close = settings.BOTANY_TOURNAMENT_CLOSE_AT
    return start < datetime.now(timezone.utc) < close
```

File: server/botany/settings.py [hunk 5]
```diff
@@ -11,7 +11,7 @@
 """
 
 import os
-from datetime import datetime, timedelta, timezone
+from datetime import datetime, timedelta, timezone, date
```
Comment (mjpieters) [inline server/botany/settings.py]: date is never used in the changeset. Make sure to run `fixup.sh` in the root, perhaps as a pre-commit hook. That'll run flake8 which would detect such issues early.

**Comment by dokato** [comment]
Thanks @mjpieters for the feedback! Please, check now.

File: server/botany/settings.py [hunk 6]
```diff
@@ -23,8 +23,20 @@
...
-# TODO read this from environment (and make optional)
-BOTANY_TOURNAMENT_CLOSE_AT = datetime(2018, 9, 18, 14, 30, tzinfo=bst)
+# Tournament time constraints in format: '%d/%m/%Y %H:%M'
+# eg. 04/11/2010 16:30.
+# If not given, +/- 1 day is a default.
+BOTANY_TOURNAMENT_START_AT = os.environ.get("BOTANY_TOURNAMENT_START_AT")
+if BOTANY_TOURNAMENT_START_AT is None:
+    BOTANY_TOURNAMENT_START_AT = datetime.now(timezone.utc) - timedelta(1)
+else:
+    BOTANY_TOURNAMENT_START_AT = datetime.strptime(BOTANY_TOURNAMENT_START_AT, '%d/%m/%Y %H:%M')
```
Comment (mjpieters) [inline server/botany/settings.py]: The parsed datetime will not have a time zone (a naive datetime) which will lead to issues later when you compare this with values that do.

Also, can I suggest that [`datetime.fromisoformat()`](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromisoformat) is used to parse the string instead? ISO8601 is a more widely understood format for this and leaves open the possibility to leave out the time, or specify seconds, and include a timezone offset, etc. 

Then, if no time zone is specified then UTC could be attached.

File: server/botany/settings.py [hunk 7]
```diff
@@ -23,8 +23,20 @@
...
+
+BOTANY_TOURNAMENT_CLOSE_AT = os.environ.get("BOTANY_TOURNAMENT_CLOSE_AT")
+if BOTANY_TOURNAMENT_CLOSE_AT is None:
+    BOTANY_TOURNAMENT_CLOSE_AT = datetime.now(timezone.utc) + timedelta(1)
+else:
+    BOTANY_TOURNAMENT_CLOSE_AT = datetime.strptime(BOTANY_TOURNAMENT_CLOSE_AT, '%d/%m/%Y %H:%M')
```
Comment (mjpieters) [inline server/botany/settings.py]: Ditto here

**Comment by mjpieters** [comment]
> I would like it if tournament_is_open was renamed back to tournament_state though, as we'll want to show different messages depending on whether the tournament is (a) yet to start; (b) running; or (c) closed.

Yeah, I was considering suggesting a`enum.Enum` state value to distinguish between upcoming, open and ended states.

**Comment by inglesp** [comment]
Hi @dokato, do you think you'll get the chance to revisit this soon?  Feel free to send me a message on Slack if you'd like to pair on any of this.

**Comment by dokato** [comment]
@inglesp I'm on it :) Just came back from a week of offline holidays.

**Comment by inglesp** [comment]
Good stuff.  No rush!

**Comment by dokato** [comment]
@inglesp @mjpieters I tried to address all your comments. Please check now.

**Comment by inglesp** [comment]
Thanks @dokato.  I've merged this in 7f13b4f34a49eda2ec071162d22f32fcef0df0bf.

I squashed the commits, and made a couple of tweaks:

* For the enums, you can use eg `TOURNAMENT_STATE.OPEN` rather than `TOURNAMENT_STATE.OPEN.value`.
* We only want to set a timezone if one is not already set.  See [here](https://github.com/inglesp/botany/blob/master/server/botany/settings.py#L33-L36).

Thanks again!

**Pull request closed without merging**